lib/utils/diagnostics/latency/monitor.go

```diff
@@ -0,0 +1,339 @@
+	for {
+		select {
+		case <-ctx.Done():
+		case then := <-pingC:
+			if err := pinger.Ping(ctx); err != nil {
+				log.WithError(err).Warn("unexpected failure sending ping")
+			} else {
+				latency.Store(m.clock.Now().Sub(then).Milliseconds())
+			}
+		}
+	}
+}
+
+// SSHClient is the subset of the [ssh.Client] required by the [SSHPinger].
+type SSHClient interface {
+	SendRequest(ctx context.Context, name string, wantReply bool, payload []byte) (bool, []byte, error)
+}
+
+// SSHPinger is a [Pinger] implementation that measure latency of an
+// SSH connection. To calculate round trip time, a keepalive@openssh.com request
+// is sent.
+type SSHPinger struct {
+	clt   SSHClient
+	clock clockwork.Clock
+}
+
+// NewSSHPinger creates a new [SSHPinger] with the provided configuration.
+func NewSSHPinger(clock clockwork.Clock, clt SSHClient) (*SSHPinger, error) {
+	if clt == nil {
+		return nil, trace.BadParameter("ssh client not provided to SSHPinger")
+	}
+
+	if clock == nil {
+		clock = clockwork.NewRealClock()
+	}
+
+	return &SSHPinger{
+		clt:   clt,
+		clock: clock,
+	}, nil
+}
+
+// Ping sends a keepalive@openssh.com request via the provided [SSHClient].
+func (s *SSHPinger) Ping(ctx context.Context) error {
+	_, _, err := s.clt.SendRequest(ctx, teleport.KeepAliveReqType, true, nil)
+	return trace.Wrap(err, "sending request keepalive@openssh.com")
+}
+
+// WebSocket is the subset of [websocket.Conn] required by the [WebSocketPinger].
+type WebSocket interface {
+	WriteControl(messageType int, data []byte, deadline time.Time) error
+	PongHandler() func(appData string) error
+	SetPongHandler(h func(appData string) error)
+}
+
+// WebSocketPinger is a [Pinger] implementation that measures the latency of a
+// websocket connection.
+type WebSocketPinger struct {
+	ws    WebSocket
+	pongC chan string
+	clock clockwork.Clock
+}
+
+// NewWebsocketPinger creates a [WebSocketPinger] with the provided configuration.
+func NewWebsocketPinger(clock clockwork.Clock, ws WebSocket) (*WebSocketPinger, error) {
+	if ws == nil {
+		return nil, trace.BadParameter("web socket not provided to WebSocketPinger")
+	}
+
+	if clock == nil {
+		clock = clockwork.NewRealClock()
+	}
+
+	pinger := &WebSocketPinger{
+		ws:    ws,
+		clock: clock,
+		pongC: make(chan string, 1),
+	}
+
+	handler := ws.PongHandler()
+	ws.SetPongHandler(func(payload string) error {
+		select {
+		case pinger.pongC <- payload:
+		default:
+		}
+
+		if handler == nil {
+			return nil
+		}
+
+		return trace.Wrap(handler(payload))
+	})
+
+	return pinger, nil
+}
+
+// Ping writes a ping control message and waits for the corresponding pong control message
+// to be received before returning. The random payload of the ping message is expected to
+// be echoed back in the pong message since it is used to determine if the pong association.
+func (s *WebSocketPinger) Ping(ctx context.Context) error {
+	// websocketPingMessage denotes a ping control message.
+	const websocketPingMessage = 9
```

Comment:
You can use gorilla's `websocket.PingMessage` (like we do for `websocket.CloseMessage` [here](https://github.com/gravitational/teleport/blob/bf2959f4b1cd85d24fbcb01c10d231c9cf89956c/lib/kube/proxy/forwarder.go#L1229))